nixos/modules/services/networking/ssh/sshd.nix

```diff
@@ -363,14 +422,9 @@ in
           HostKey ${k.path}
         '')}
 
-        ### Recommended settings from both:
-        # https://stribika.github.io/2015/01/04/secure-secure-shell.html
-        # and
-        # https://wiki.mozilla.org/Security/Guidelines/OpenSSH#Modern_.28OpenSSH_6.7.2B.29
-
-        KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
-        Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
-        MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com
+        KexAlgorithms ${concatStringsSep "," cfg.kexAlgorithms}
+        Ciphers ${concatStringsSep "," cfg.ciphers}
+        MACs ${concatStringsSep "," cfg.macs}
```

Comment:
and then only write these out if the options are specified.

This way, we don't have to maintain our own default list which will inevitably become out of sync with sshd's recommended options.

Reply:
We do need a default, we can't just have everybody specify what algs to use in their config all the time. You couldn't just `services.openssh.enable = true` and have a working openssh anymore.

Reply:
They won't need to:

```
       KexAlgorithms
              Specifies the available  KEX  (Key  Exchange)  algo‐
              rithms.   Multiple  algorithms  must  be comma-sepa‐
              rated.  Alternately if the  specified  value  begins
              with  a  `+'  character,  then the specified methods
              will be appended  to  the  default  set  instead  of
              replacing  them.  If the specified value begins with
              a `-' character, then the specified methods (includ‐
              ing  wildcards) will be removed from the default set
              instead of replacing them.  The default is:

              curve25519-sha256,curve25519-sha256@libssh.org,
              ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-
              sha2-nistp521,
              diffie-hellman-group-exchange-sha256,
              diffie-hellman-group-exchange-sha1,
              diffie-hellman-group14-sha1

              The  list  of  available key exchange algorithms may
              also be obtained using Qq ssh -Q kex .

       Ciphers
              Specifies the ciphers allowed  and  their  order  of
              preference.   Multiple  ciphers  must be comma-sepa‐
              rated.  If the specified value  begins  with  a  `+'
              character,   then  the  specified  ciphers  will  be
              appended to the default  set  instead  of  replacing
              them.   If  the  specified  value  begins with a `-'
              character, then  the  specified  ciphers  (including
              wildcards)  will  be  removed  from  the default set
              instead of replacing them.

              The supported ciphers are:

              3des-cbc
              aes128-cbc
              aes192-cbc
              aes256-cbc
              aes128-ctr
              aes192-ctr
              aes256-ctr
              aes128-gcm@openssh.com
              aes256-gcm@openssh.com
              chacha20-poly1305@openssh.com

              The default is:

              chacha20-poly1305@openssh.com,
              aes128-ctr,aes192-ctr,aes256-ctr,
              aes128-gcm@openssh.com,aes256-gcm@openssh.com,
              aes128-cbc,aes192-cbc,aes256-cbc

       MACs   Specifies  the  MAC  (message  authentication  code)
              algorithms in order of preference.   The  MAC  algo‐
              rithm is used for data integrity protection.  Multi‐
              ple algorithms  must  be  comma-separated.   If  the
              specified  value  begins  with a `+' character, then
              the specified algorithms will  be  appended  to  the
              default set instead of replacing them.  If the spec‐
              ified value begins with a `-'  character,  then  the
              specified  algorithms  (including wildcards) will be
              removed from the default set  instead  of  replacing
              them.

              The  algorithms  that  contain Qq -etm calculate the
              MAC after encryption (encrypt-then-mac).  These  are
              considered safer and their use recommended.

              The default is:

              umac-64-etm@openssh.com,umac-128-etm@openssh.com,
              hmac-sha2-256-etm@openssh.com,hmac-
              sha2-512-etm@openssh.com,
              hmac-sha1-etm@openssh.com,
              umac-64@openssh.com,umac-128@openssh.com,
              hmac-sha2-256,hmac-sha2-512,hmac-sha1

              The list of available MAC  algorithms  may  also  be
              obtained using Qq ssh -Q mac .

```

Reply:
These are insecure defaults, see https://stribika.github.io/2015/01/04/secure-secure-shell.html (linked from the options descriptions), and the PR that added such defaults originally: https://github.com/NixOS/nixpkgs/pull/31763

Reply:
cc @andir and @fpletz for an opinion here

Reply:
The way I see it, these should either be provided by default, as they care now, or as a secure profile, which was the original discussion point which sparked me to originally do the work. However,

1. We should maintain these settings in _some_ profile, default or otherwise.
2. Should the secure settings not be the default settings?

Yes, this is additional work for NixOS security maintenance, but it's better for the security of every  NixOS user, by default, and it doesn't make anything harder for them.

Reply:
My opinion is if they are in any profile it should be The Default.

Reply:
I see the point of the ageing cipher selection and having to maintain them. As @grahamc said we will "inevitably become out of sync" with whatever options one might deem as the holy grail of secure defaults.

BUT I also would argue that we need sane defaults. The defaults supplied by OpenSSH are probably not was many people would prefer.

I agree with Graham that if we have a "secure" profile that should be the default. If we had a "paranoid" profile that could be opt-in. If you have to downgrade your settings then you would disable the profile (`services.openssh.server.profiles.default.enable = false`, or some better option) and/or just define your custom list of algorithms.

How about we attempt to remove old (and "insecure") algorithms instead with the `-` qualifier? That would potentially allow new ciphers (which should be more secure?!?) to be picked up while we can easily disable ciphers as required.

I am not really happy with my proposal either. I would really like that the people dealing with the OpenSSH packaging would also maintain this list. They should come across changes in ciphersuites while reading the change logs anyway. The problem there is that it is yet another burden for a small number of people...

Reply:
(Maybe stupid) idea: have some tag we can grep for in the source code saying “This will need to be reviewed before XXXX-XX-XX and every Y months/years”, and have someone setup a cron to automatically warn when XXXX-XX-XX is passed without this tag being updated?

This would simplify maintenance of such lists, for which I guess OpenSSH wouldn't be the sole user. And I don't think there is a big hurry for upgrade (or else everyone will hear about it): from my limited experience, when a cipher breaks, it breaks slowly over time, and not in a few months. Like, everyone have known for a while that SHA1 is not really secure, yet it still stays relatively OK for quite a lot of use cases.